src/psyclone/tests/psyir/frontend/fparser2_derived_type_test.py

```diff
@@ -0,0 +1,154 @@
+                                 "type(my_type) :: var")
+    fparser2spec = Specification_Part(reader)
+    processor.process_declarations(fake_parent, fparser2spec.content, [])
+    vsym = symtab.lookup("var")
+    assert isinstance(vsym.datatype, DeferredType)
+    tsym = symtab.lookup("my_type")
+    assert isinstance(tsym, TypeSymbol)
+
+
+@pytest.mark.usefixtures("f2008_parser")
+def test_missing_derived_type():
+    ''' Check that the fronted raises an error if it encounters a variable
+    of a derived type that cannot be resolved. '''
+    fake_parent = KernelSchedule("dummy_schedule")
+    processor = Fparser2Reader()
+    reader = FortranStringReader("type(my_type) :: var")
+    fparser2spec = Specification_Part(reader)
+    # This should raise an error because there's no Container from which
+    # the definition of 'my_type' can be brought into scope.
+    with pytest.raises(SymbolError) as err:
+        processor.process_declarations(fake_parent, fparser2spec.content, [])
+    assert "No Symbol found for name 'my_type'" in str(err.value)
+
+
+def test_name_clash_derived_type(f2008_parser):
+    ''' Check that the fronted raises an error if it encounters a reference
+    to a derived type that clashes with another symbol. '''
+    fake_parent = KernelSchedule("dummy_schedule")
+    symtab = fake_parent.symbol_table
+    # Add a RoutineSymbol to the symbol table that clashes with the name
+    # of the derived type.
+    symtab.add(RoutineSymbol("my_type"))
+    processor = Fparser2Reader()
+    reader = FortranStringReader("subroutine my_sub()\n"
+                                 "  type(my_type) :: some_var\n"
+                                 "end subroutine my_sub\n")
+    fparser2spec = f2008_parser(reader)
+    # This should raise an error because the Container symbol table should
+    # already contain a RoutineSymbol named 'my_type'
+    with pytest.raises(SymbolError) as err:
+        processor.process_declarations(fake_parent, fparser2spec.content, [])
+    assert ("Search for a TypeSymbol named 'my_type' found a RoutineSymbol "
+            "instead" in str(err.value))
+
+
+@pytest.mark.usefixtures("f2008_parser")
+@pytest.mark.parametrize("use_stmt", ["use grid_mod, only: grid_type",
+                                      "use grid_mod"])
+def test_parse_derived_type(use_stmt):
+    ''' Check that the fronted correctly creates a TypeSymbol of type
+    StructureType from the declaration of a derived type. '''
+    fake_parent = KernelSchedule("dummy_schedule")
+    symtab = fake_parent.symbol_table
+    processor = Fparser2Reader()
+    reader = FortranStringReader("{0}\n"
+                                 "type :: my_type\n"
+                                 "  integer :: flag\n"
+                                 "  type(grid_type), private :: grid\n"
+                                 "  real, dimension(3) :: posn\n"
+                                 "end type my_type\n".format(use_stmt))
```

Comment:
Yes although that then demonstrated that I still had some bugs. I've re-ordered the code in the implementation so that it checks for derived-type definitions *before* it looks at the variable declarations. The code was also expecting the type of the derived-type symbol to be (a copy of) the type of the associated type symbol. I've changed this so that it now expects it to *be* the type symbol.